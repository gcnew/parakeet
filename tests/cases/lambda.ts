
import { test } from '../test_util'

import {
    char, oneOf,

    CharNotExpected, StringMismatch,

    TextStream, StringParser
} from '../../src/string_combinators'

import {
    map, alt, alt3, many, combine, combine3, combine4, terminated,

    satisfy as psat,
    oneOrMore as many1
} from '../../src/parser_combinators'


/* Types */
type Literal     = { kind: 'literal',     value: number }
type Reference   = { kind: 'reference',   ref:   string }
type Application = { kind: 'application', func:  Expr,   arg:  Expr }
type Lambda      = { kind: 'lambda',      param: string, body: Expr }

type Expr = Literal | Reference | Application | Lambda


/* Constructors */
function literal(value: number): Literal {
    return { kind: 'literal', value };
}

function reference(ref: string): Reference {
    return { kind: 'reference', ref };
}

function application(func: Expr, arg: Expr): Application {
    return { kind: 'application', func, arg };
}

function lambda(param: string, body: Expr): Lambda {
    return { kind: 'lambda', param, body };
}


/* Parsers */
const parseWs    = many1(oneOf(' \t'));
const parseNum   = many1(oneOf('1234567890'));

const parseId    = alt(
    oneOf('+-*/⊥'),
    map(
        many1(psat<char, TextStream, CharNotExpected>(
            x => /[a-zA-Z]/.test(x),
            { kind: 'pc_error', code: 'char_not_expected', expected: '[a-zA-Z]' }
        )),
        parts => parts.join('')
    )
);

// Lambda = 'λ' Id '.' Expr
const parseLambda = combine4(
    char('λ'),
    parseId,
    char('.'),
    st => parseExpr(st),
    (_1, param, _2, body) => lambda(param, body)
);

// ExprNoAp = Literal
//          | Ref           -- ref <$> id
//          | '(' Expr ')'
const parseExprNoAp = alt3(
    map(parseNum, n   => literal(+n.join(''))),
    map(parseId,  ref => reference(ref)),
    combine3(char('('), st => parseExpr(st), char(')'), (_1, e, _2) => e)
);

// Expr = Lambda
//      | ExprNoAp (' ' ExprNoAp)*
const parseExpr: StringParser<StringMismatch, Expr> = alt(
    parseLambda,
    combine(
        parseExprNoAp,
        many(combine(parseWs, parseExprNoAp, (_, expr) => expr)),
        (head, rest) => rest.reduce(application, head)
    )
);

const parseProgram = terminated(parseExpr);

test(parseProgram, 'λf.f (λx.x) + 1');
test(parseProgram, '(λf.λx.f x) + 1');
test(parseProgram, 'λf.f λx.x + 1');
test(parseProgram, 'λx.λxs.λcc.λcn.cc x xs');
test(parseProgram, 'λtrue.λfalse.true (true false)');
