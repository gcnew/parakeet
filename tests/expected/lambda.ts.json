{
    "kind": "right",
    "value": [
        {
            "kind": "lambda",
            "param": "f",
            "body": {
                "kind": "application",
                "func": {
                    "kind": "application",
                    "func": {
                        "kind": "application",
                        "func": {
                            "kind": "reference",
                            "ref": "f"
                        },
                        "arg": {
                            "kind": "lambda",
                            "param": "x",
                            "body": {
                                "kind": "reference",
                                "ref": "x"
                            }
                        }
                    },
                    "arg": {
                        "kind": "reference",
                        "ref": "+"
                    }
                },
                "arg": {
                    "kind": "literal",
                    "value": 1
                }
            }
        },
        {
            "source": "λf.f (λx.x) + 1",
            "offset": 15,
            "data": [
                [
                    0,
                    "λf.f (λx.x) + 1"
                ]
            ]
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "application",
            "func": {
                "kind": "application",
                "func": {
                    "kind": "lambda",
                    "param": "f",
                    "body": {
                        "kind": "lambda",
                        "param": "x",
                        "body": {
                            "kind": "application",
                            "func": {
                                "kind": "reference",
                                "ref": "f"
                            },
                            "arg": {
                                "kind": "reference",
                                "ref": "x"
                            }
                        }
                    }
                },
                "arg": {
                    "kind": "reference",
                    "ref": "+"
                }
            },
            "arg": {
                "kind": "literal",
                "value": 1
            }
        },
        {
            "source": "(λf.λx.f x) + 1",
            "offset": 15,
            "data": [
                [
                    0,
                    "(λf.λx.f x) + 1"
                ]
            ]
        }
    ]
}

{
    "kind": "left",
    "value": {
        "kind": "pc_error",
        "code": "eos_expected"
    }
}

{
    "kind": "right",
    "value": [
        {
            "kind": "lambda",
            "param": "x",
            "body": {
                "kind": "lambda",
                "param": "xs",
                "body": {
                    "kind": "lambda",
                    "param": "cc",
                    "body": {
                        "kind": "lambda",
                        "param": "cn",
                        "body": {
                            "kind": "application",
                            "func": {
                                "kind": "application",
                                "func": {
                                    "kind": "reference",
                                    "ref": "cc"
                                },
                                "arg": {
                                    "kind": "reference",
                                    "ref": "x"
                                }
                            },
                            "arg": {
                                "kind": "reference",
                                "ref": "xs"
                            }
                        }
                    }
                }
            }
        },
        {
            "source": "λx.λxs.λcc.λcn.cc x xs",
            "offset": 22,
            "data": [
                [
                    0,
                    "λx.λxs.λcc.λcn.cc x xs"
                ]
            ]
        }
    ]
}

{
    "kind": "right",
    "value": [
        {
            "kind": "lambda",
            "param": "true",
            "body": {
                "kind": "lambda",
                "param": "false",
                "body": {
                    "kind": "application",
                    "func": {
                        "kind": "reference",
                        "ref": "true"
                    },
                    "arg": {
                        "kind": "application",
                        "func": {
                            "kind": "reference",
                            "ref": "true"
                        },
                        "arg": {
                            "kind": "reference",
                            "ref": "false"
                        }
                    }
                }
            }
        },
        {
            "source": "λtrue.λfalse.true (true false)",
            "offset": 30,
            "data": [
                [
                    0,
                    "λtrue.λfalse.true (true false)"
                ]
            ]
        }
    ]
}
